pkg/boot/crawl/crawl.go

```diff
@@ -2,132 +2,233 @@ package crawl
-}
+var ErrClosed = errors.New("closed")
 
-type Scanner interface {
-	Scan(net.Conn, record.Record) (*record.Envelope, error)
+type advRequest struct {
+	ns  string
+	ttl time.Duration
 }
 
 type Crawler struct {
-	Logger   log.Logger
-	Dialer   Dialer
-	Strategy DialStrategy
-	Scanner  Scanner
-}
+	cidr string
+	port int
 
-func (c Crawler) FindPeers(ctx context.Context, ns string, opt ...discovery.Option) (<-chan peer.AddrInfo, error) {
-	if c.Logger == nil {
-		c.Logger = log.New(log.WithLevel(log.FatalLevel))
-	}
+	done  <-chan struct{}
+	cherr chan<- error
+	err   atomic.Value
 
-	if c.Scanner == nil {
-		c.Scanner = basicScanner{}
-	}
+	t              time.Time
+	advertisements chan<- advRequest
+	mustAdvertise  map[string]time.Time
 
-	c.Logger = c.Logger.WithField("ns", ns)
+	rec atomic.Value
+}
 
-	var opts discovery.Options
-	if err := opts.Apply(opt...); err != nil {
+func New(h host.Host, cidr string, port int) (*Crawler, error) {
+	var (
+		cherr          = make(chan error, 1)
+		done           = make(chan struct{})
+		advertisements = make(chan advRequest)
+		requests       = make(chan request)
+	)
+
+	c := &Crawler{
+		cidr:           cidr,
+		port:           port,
+		done:           done,
+		t:              netutil.Time(),
+		advertisements: advertisements,
+		mustAdvertise:  make(map[string]time.Time),
+		cherr:          cherr,
+	}
+
+	if err := c.trackHostAddr(h); err != nil {
 		return nil, err
 	}
 
-	conns, err := c.Strategy.Dial(ctx, c.dialer())
+	comm, err := newCommFromCIDR(cidr, port)
 	if err != nil {
 		return nil, err
 	}
 
-	out := make(chan peer.AddrInfo, 8)
 	go func() {
-		defer close(out)
-
-		c.Logger.Debug("crawl started")
-		defer c.Logger.Debugf("crawl finished")
-
-		var rec peer.PeerRecord
-		for conn := range conns {
-			if c.read(ctx, conn, &rec) {
-				select {
-				case out <- peer.AddrInfo{ID: rec.PeerID, Addrs: rec.Addrs}:
-					if opts.Limit--; opts.Limit == 0 {
-						return
-					}
+		defer close(done)
+		defer comm.close()
+
+		go comm.receiveRequests(requests)
 
-				case <-ctx.Done():
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case t := <-ticker.C:
+				for ns, deadline := range c.mustAdvertise {
+					if t.After(deadline) {
+						delete(c.mustAdvertise, ns)
+					}
 				}
+			case adv := <-advertisements:
+				c.mustAdvertise[adv.ns] = c.t.Add(adv.ttl)
+			case r := <-requests:
+				if _, ok := c.mustAdvertise[r.ns]; ok {
+					envelope, _ := c.rec.Load().(*record.Envelope).Marshal()
+					go comm.sendTo(envelope, r.addr)
+				}
+			case err := <-cherr:
+				c.err.CompareAndSwap(error(nil), err)
+				return
 			}
 		}
 	}()
 
-	return out, nil
+	return c, nil
 }
 
-func (c Crawler) read(ctx context.Context, conn net.Conn, r record.Record) bool {
-	defer conn.Close()
-
-	if err := c.deadline(ctx, conn); err != nil {
-		c.Logger.WithError(err).Debug("unable to set deadline")
-		return false
+func (c *Crawler) trackHostAddr(h host.Host) error {
```

Comment:
This will block indefinitely for client hosts.  Can we set this up lazily on the first call to `Advertise`, e.g. using `sync.Once` ?  We should also have some kind of best-effort check that the host is actually listening for connections (e.g. `len(h.Addrs()) > 0`) in order to catch user mistakes.